lib/broccoli/service-worker-manifest.js

```diff
@@ -0,0 +1,115 @@
+var diffing_broccoli_plugin_1 = require('./diffing-broccoli-plugin');
+var path = require('path');
+var fs = require('fs');
+var crypto = require('crypto');
+var FILE_ENCODING = { encoding: 'utf-8' };
+var MANIFEST_FILE = 'manifest.appcache';
+var FILE_HASH_PREFIX = '# sw.file.hash:';
+
+var DiffingSWManifest = (function () {
+  function DiffingSWManifest(inputPath, cachePath, options) {
+    this.inputPath = inputPath;
+    this.cachePath = cachePath;
+    this.options = options;
+    this.firstBuild = true;
+  }
+  // Called by the diffing broccoli plugin with a delta of files to process.
+  DiffingSWManifest.prototype.rebuild = function(diff) {
+    var _this = this;
+    var manifest = {};
+    if (this.firstBuild) {
+      this.firstBuild = false;
+    } else {
+      // Read manifest from disk.
+      manifest = this.readManifestFromCache();
+    }
+    
+    // Remove manifest entries for files that are no longer present.
+    diff.removedPaths.forEach(function(file) {
+      delete manifest[file];
+    });
+    
+    // Merge the lists of added and changed paths and update their hashes in the manifest.
+    []
+      .concat(diff.addedPaths)
+      .concat(diff.changedPaths)
+      .filter(function(file) {
+        // Exclude the manifest file, since it may be counted as an add.
+        return file !== MANIFEST_FILE;
+      })
+      .forEach(function(file) {
+        manifest[file] = _this.computeFileHash(file);
+      });
+    var manifestPath = path.join(this.cachePath, MANIFEST_FILE);
+    fs.writeFileSync(manifestPath, this.generateManifest(manifest));
+  };
+  // Compute the hash of the given relative file.
+  DiffingSWManifest.prototype.computeFileHash = function(file) {
+    var contents = fs.readFileSync(path.join(this.inputPath, file));
+    return crypto
+      .createHash('sha1')
+      .update(contents)
+      .digest('hex');
+  };
+  // Compute the hash of the bundle from the names and hashes of all included files.
+  DiffingSWManifest.prototype.computeBundleHash = function(files, manifest) {
+    var hash = crypto.createHash('sha1');
+    files.forEach(function(file) {
+      // Hash comes first to avoid any weird issues with filenames.
+      hash.update(manifest[file] + ':' + file);
+    });
+    return hash.digest('hex');
+  };
+  // Generate the string contents of the manifest.
+  DiffingSWManifest.prototype.generateManifest = function(manifest) {
+    var _this = this;
+    var files = [];
+    for (var file in manifest) {
+      files.push(file);
+    }
+    // Put files in alphabetical order to ensure consistency between builds.
+    files.sort();
+    
+    var contents = "CACHE MANIFEST\n# sw.bundle: ng-cli\n# sw.version: ";
```

Comment:
you can use quasi-literals templates now in Node. This may really cut your code in half and makes it easier to read.